### ISSUE TITLE
OOM recovery under multi-node FSDP/HSDP

### ISSUE THREAD

**Issue by vadimkantorov** [pytorch/torchtitan#1329]
### Bug description

Does torchtitan provide any recipes of how to implement batch skipping / OOM recovery in multi-node FSDP setup?

In RL/GRPO training this is very pertinent (where we don't know response seqlens a-priori to do packing / clipping):
- https://github.com/volcengine/verl/issues/2159

One thing I could think of:
- some sort of micro-batching for backward pass
- some generic batch skipping

Some sort of memory operation tracing would also be very useful to better know what is the reason of OOM (fragmentation):
- https://github.com/pytorch/pytorch/issues/91692#issuecomment-2996838221

### Versions

N/A

**Comment by fegin** [comment]
Is the question about recovery from OOM or debugging an OOM? There are two tools that can help you to debug OOM:
1. `--profiling.enable_memory_snapshot` which will give you the memory for debugging, https://pytorch.org/blog/understanding-gpu-memory-1/.
2. Another possibility is that some tensors are not freed because of GC being disabled and these tensors are involved in circular dependencies that require GC to recycle. `--training.gc_debug` can help you debug this issue. If there are tensor involved in circular dependencies, you will see a warning when this flag is on.

If your question is regarding recovery, afaik, you can only restart the trainer as some computation and communication kernels may be in an undefined states.

**Comment by vadimkantorov** [comment]
Mainly the question is regarding recovery.

I saw at some point some workarounds like https://github.com/facebookresearch/ParlAI/pull/2704

I wonder if any of these hacks can reliably be upgraded to the modern FSDP/HSDP. E.g. in the verl setup, we're having OOM crash during backward. We're looking for a way that would simply ignore/skip the batches if OOM occurs during forward or backward. Is it possible to somehow reset the gradients / still satisfy the communication collectives, but at the end somehow skip doing the optimizer step?

For DDP, there were comments like https://discuss.pytorch.org/t/handeling-oom-in-backward-pass-using-ddp/103916/5 : 
`+1 to [@osalpekar](https://discuss.pytorch.org/u/osalpekar)’s comment that [torchelastic](https://pytorch.org/elastic) is the recommended solution for DDP OOMs. The RuntimeError you saw is caused by desync, and the desync is caused by OOM in one process. Because DDP expects all processes to launch the same number of AllReduce comm ops in every iteration. If one process hit OOM and skip/redo some comm ops, it will break this assumption. TorchElastic handles this problem by checkpointing model states and let the entire DDP gang to revert to the previous checkpoint when it detects failures in any process, which seems to be very nice fit for this infrequent and unpredictable OOM.`

But not sure, how this got transformed with torchelastic upstreaming to PyTorch. Can't find any information about this, and if this feature can be upgraded to FSDP

This problem is more acute for GRPO/RL training when we don't know response seqlen a-priori and can't do seqlen binning/packing... Having a PyTorch recipe/reliable building block for skipping OOM-d batches for FSDP is needed by many frameworks.

---

For debugging, yes, it would be extremely good to have breakdown like in the blog you linked (maybe in addition to existing memory stats dict in pytorch - also maybe adding gradient checkpoints, and any tensors saved_for_backward, and also having some measure of memory fragmentation would be good to have an idea if it's a problem or not):

![](https://pytorch.org/wp-content/uploads/2024/11/fig6-2.png)

**Comment by fegin** [comment]
The `torchelastic` comment is the same as my original comment -- "restart the trainer as some computation and communication kernels may be in an undefined states.". You can think `TorchElastic` as a launcher. So using `TorchElastic` means that `TorchElastic` restarts the trainers to avoid the deadlock. FSDP is no exception, as there are allgather/reduce_scater. Once OOM happens, there can be a hang and a timeout eventually. We cannot reliably recover from a timeout.

The technique you are looking into is TorchFT, which is integrated with TorchTitan's HSDP, which the failing group (OOM group) will be restarted and the rest groups will continue the training, you can check the most recently post: https://fburl.com/6t3kp1jh. Note that the failing group, not just the failing node will be kicked out from the training until the whole group recover, so this may not be the best solution for this case.

**Comment by d4l3k** [comment]
I think torchft could help here though OOMs seem like they should be solvable via application code. 

@vadimkantorov are simple max seqlen settings not sufficient to work around this issue? Or you want something that requires no tuning?

**Comment by vadimkantorov** [comment]
> are simple max seqlen settings not sufficient to work around this issue?

The problem becomes when in a batch with variable completion lens (unknown a-priori), solving this with max seqlen would lead to reduction of batch size.

In the RL setting, some workarounds like packed microbatching might be possible. But there is also a need of having a building block/recipe for some last resort OOM crash protection which would prevent training run crashes. At least a plug-in recipe for some common cases / models (like the hack in ParlAI or a similar hack in detectron2)

**Comment by tianyu-l** [comment]
trying to better understand the use case..

> The problem becomes when in a batch with variable completion lens (unknown a-priori), solving this with max seqlen would lead to reduction of batch size.

Assuming packing is used, what if you start with fixed seq len and fixed local batch size -- if you'd require larger global batch size due to unbounded completion lens, you could use gradient accumulation.

**Comment by vadimkantorov** [comment]
@fegin Would you know, what is AUTOGRAD_DETAIL?

My actual problem is how to integrate memory introspection into the verl framework. And how to backport TorchFT/torchelastic or any other tools to verl so that it can skip/recover from OOMs

**Comment by vadimkantorov** [comment]
Another issue preventing OOM recovery / continuing after OOM:
- https://github.com/pytorch/pytorch/issues/82218#issuecomment-1633060314
